src/rpc/pd/client.rs

```diff
@@ -0,0 +1,218 @@
+// Copyright 2018 The TiKV Project Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::{
+    fmt,
+    sync::{Arc, RwLock},
+    time::Duration,
+};
+
+use futures::Future;
+use grpcio::{CallOption, Environment};
+use kvproto::{metapb, pdpb, pdpb_grpc::PdClient as RpcClient};
+
+use crate::{
+    rpc::{
+        pd::{
+            context::{request_context, PdRequestContext},
+            leader::LeaderClient,
+            request::Request,
+            PdTimestamp, Region, RegionId, Store, StoreId,
+        },
+        security::SecurityManager,
+        util::HandyRwLock,
+    },
+    Error, Result,
+};
+
+const LEADER_CHANGE_RETRY: usize = 10;
+
+trait PdResponse {
+    fn header(&self) -> &pdpb::ResponseHeader;
+}
+
+macro_rules! pd_response {
```

Comment:
This macro doesn't save space. 🤣 I think it's ok though.

Reply:
Hm, looks like there are only couple of types a client eventually has to handle. So it really doesn't save much. Let's remove this macro and make it easier to read.